src/centralized_telescope.rs

```diff
@@ -0,0 +1,421 @@
+                    if round.h_u64 == h {
+                        if let Some(r) = Round::update(&round, element, setup.sec_param) {
+                            round = r;
+                        } else {
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return Self::h2(setup, &round);
+        }
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::gen_items;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_verify() {
+        let mut rng = ChaCha20Rng::from_seed(Default::default());
+        let nb_tests = 1_000;
+        let set_size = 1_000;
+        for _t in 0..nb_tests {
+            let seed = rng.next_u32().to_be_bytes().to_vec();
+            let s_p = gen_items::<DATA_LENGTH>(&seed, set_size);
+            let params = Params {
+                lambda_sec: 10.0,
+                lambda_rel: 10.0,
+                n_p: 80,
+                n_f: 20,
+            };
+            let setup = Setup::new(&params);
+            let proof = Proof::prove(&setup, &s_p).unwrap();
+            assert!(Proof::verify(&setup, &proof.clone()));
+            let proof_d = Proof {
+                v: proof.v,
+                t: proof.t.wrapping_add(1),
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, &proof_d));
+            let proof_r = Proof {
+                v: proof.v.wrapping_add(1),
+                t: proof.t,
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, &proof_r));
+            let proof_item = Proof {
+                v: proof.v,
+                t: proof.t,
+                items: Vec::new(),
+            };
+            assert!(!Proof::verify(&setup, &proof_item));
+            let mut wrong_items = proof.items.clone();
+            let last_item = wrong_items.pop().unwrap();
+            let mut penultimate_item = wrong_items.pop().unwrap();
+            let proof_itembis = Proof {
+                v: proof.v,
+                t: proof.t,
+                items: wrong_items.clone(),
+            };
+            assert!(!Proof::verify(&setup, &proof_itembis));
+            // Modifying the penultimate item to check correctness of H1 check and not H2
+            penultimate_item[0] = penultimate_item[0].wrapping_add(42u8);
+            wrong_items.push(penultimate_item);
+            wrong_items.push(last_item);
+            let proof_itembis = Proof {
+                v: proof.v,
+                t: proof.t,
+                items: wrong_items.clone(),
+            };
+            assert!(!Proof::verify(&setup, &proof_itembis));
+        }
+    }
+}
```

Comment:
Do we have an issue for (basic) tests? I think we could benefit from testing the DFS bound logic, parameter derivation, and maybe some other tests.

Reply:
Sure, feel free to add more tests in a separate issue: https://github.com/cardano-scaling/alba/issues/49